### PR TITLE
[WIP] Add paddle support

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
Support Paddle framework in array-api-compat, but still working in progress.

TODO List:
- [x] https://github.com/PaddlePaddle/Paddle/pull/69632
- [x] https://github.com/PaddlePaddle/Paddle/pull/69477

Related issue: <https://github.com/PaddlePaddle/Paddle/issues/68618>

### CODE REVIEW

**Comment by rgommers** [comment]
Cool, thanks for working on this @HydrogenSulfate!

I am curious to learn a bit more about Paddle. In particular conceptually what is supported - https://www.paddlepaddle.org.cn/documentation/docs/en/guides/jit/index_en.html and a few other guides tell me a bit, but not quite what I was most interested in. A few questions if you don't mind:
- Is the default execution model eager or lazy/graph-based?
- It looks like there is a JIT compiler, what's the syntax and does it work similar to, for example, `jax.jit` or `torch.compile`?
- Is item and slice assignment supported via `__setitem__`? And indexing with boolean mask?
- Is mixed integer and floating-point type promotion supported?
- I see it has CPU and NVIDIA GPU support, plus some other vendors of accelerators that I don't immediately recognize. Are those all GPUs as well? And ROCm and Intel XPUs are not supported (now or in the near future)?

**Comment by HydrogenSulfate** [comment]
> Cool, thanks for working on this @HydrogenSulfate!
> 
> I am curious to learn a bit more about Paddle. In particular conceptually what is supported - https://www.paddlepaddle.org.cn/documentation/docs/en/guides/jit/index_en.html and a few other guides tell me a bit, but not quite what I was most interested in. A few questions if you don't mind:
> 
> * Is the default execution model eager or lazy/graph-based?
> * It looks like there is a JIT compiler, what's the syntax and does it work similar to, for example, `jax.jit` or `torch.compile`?
> * Is item and slice assignment supported via `__setitem__`? And indexing with boolean mask?
> * Is mixed integer and floating-point type promotion supported?
> * I see it has CPU and NVIDIA GPU support, plus some other vendors of accelerators that I don't immediately recognize. Are those all GPUs as well? And ROCm and Intel XPUs are not supported (now or in the near future)?

Thanks much for reply and attention to this PR,
1. Paddle use eager execution mode as default(eager Tensor running with dynamic graph), and can be manually switched to static graph(lazy Tensor running with static computational Program) by `model = paddle.jit.to_static(mode)`.
2. The usage of `paddle.jit.to_static` and `torch.compile`/`jiax.jit` is very similar. When designing these interfaces, we referred to influential and great tools such as pytorch/jax. The usage of `paddle.jit.*` is roughly as follows:


    1. Firstly, users will use dynamic graphs for programming and training models
    2. Secondly, if users need it, they can use one line of code to convert the model: `model=paddle. jit.to_static(model)` without any other modifications, convert the model to a static graph model, and then start training. Due to the advantages of static graph models, this usually results in a small performance improvement, and our conversion rate has been extensively tested on our existing models, with a success rate close to 100%
    3. If there is a higher performance requirement, users can add the option to enable the CINN compiler in `jit.to_static`: [modulus-sym code for exmample](< https://github.com/PaddlePaddle/modulus-sym/blob/45d576918f26f1ae132cf075534f8606e8d96090/modulus/sym/domain/constraint/constraint.py#L107 -L115>), which can capture the entire computation graph, including forward pass, backward pass, even double-backward pass(or higher-order), and further accelerate the program. We have tested it on 40+models in the NVIDIA/modulus-sym suite and achieved IPS performance that exceeds PyTorch by [about 70%](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/paddle_v3_features/higher_order_ad_cn.html#ai-for-science) when the [CINN compiler](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/paddle_v3_features/cinn_cn.html) is enabled (of course, this is partly because PyTorch does not seem to support capturing and compiling high-order reverse graphs)
 
3. item and slice assignment are supported with broadcasting as below
    ``` py
    import paddle
    
    x = paddle.randn([4, 3, 2])
    v = paddle.randn([3, 2])
    x[:] = v
    print(x)
    
    mask = paddle.to_tensor([True, False, True, False])
    
    x[mask] = paddle.zeros([3, 2])
    print(x)
    ```
4. Our implicit promotion support fp32/fp64, c32/c64 promotion, but do not support mixed integer and bool type(the purpose is to avoid covert transformations that are easily overlooked by users, which can lead to the model giving unexpected results), detailed table can be checked at [url](https://www.paddlepaddle.org.cn/documentation/docs/zh/guides/advanced/auto_type_promotion_cn.html#sifeijiangyinshileixingtishengdeshiyongfanwei): 
image

5. We support [XPU](https://www.paddlepaddle.org.cn/documentation/docs/zh/hardware_support/xpu/index_cn.html#xpu) and [ROCM](https://www.paddlepaddle.org.cn/documentation/docs/zh/dev_guides/sugon/paddle_c86_cn.html#rocm), I will supplement these devices type in subsequent commits

**Comment by rgommers** [comment]
Thanks for the detailed explanation @HydrogenSulfate, much appreciated.


> 5\. We support [XPU](https://www.paddlepaddle.org.cn/documentation/docs/zh/hardware_support/xpu/index_cn.html#xpu) and [ROCM](https://www.paddlepaddle.org.cn/documentation/docs/zh/dev_guides/sugon/paddle_c86_cn.html#rocm), I will supplement these devices type in subsequent commits

I'll note that I tried inferring supported devices from [this Install page](https://www.paddlepaddle.org.cn/en/install/quick?docurl=undefined), where ROCm/XPU aren't yet present:

image

**Comment by HydrogenSulfate** [comment]
> 谢谢你的详细解释@HydrogenSulfate，非常感谢。
> 
> > 5. 我们支持[XPU](https://www.paddlepaddle.org.cn/documentation/docs/zh/hardware_support/xpu/index_cn.html#xpu)和[ROCM](https://www.paddlepaddle.org.cn/documentation/docs/zh/dev_guides/sugon/paddle_c86_cn.html#rocm)，我会在后续的提交中补充这些设备类型
> 
> 我会注意到，我尝试从此[安装页面](https://www.paddlepaddle.org.cn/en/install/quick?docurl=undefined)推断支持的设备，但其中 ROCm/XPU 尚不存在：
> 
> 图像

Embarrassingly, our English documents are somewhat outdated, so you can use the browser's translation feature to translate the Chinese documents into English.

ROCm is used in [`HYGON`](https://www.paddlepaddle.org.cn/documentation/docs/zh/guides/hardware_support/dcu/install_cn.html): 
image
image

XPU is used in [`KUNLUNXIN`](https://www.paddlepaddle.org.cn/documentation/docs/zh/install/install_xpu_cn.html):
image

**Comment by rgommers** [comment]
> Embarrassingly, our English documents are somewhat outdated

Not embarrassing at all - we still haven't even deployed our Chinese translations on https://numpy.org/ (they're coming though!).

Thanks for the tips. Once this is ready, I'll try giving Paddle + SciPy a spin.

**Comment by lucascolley** [comment]
I converted this PR to draft since it looks like it is still a work in progress @HydrogenSulfate , but feel free to let us know whenever it is ready for another look!
